File: duplicateKeyValue.js
// description: Check if there is no duplicate key value with a data set
// duplicateKeyValue.js
//
// Inputs are: the keys list to search for duplicates
//    Input type: an object arg containing a string
// Outputs are: the list of keys which are duplicated
//    Output type: config datasets failing the validation
//
// Creator: Cyrille
// Maintainer:
// Version:   0.9
// Support: Sweagle version >= 3.11

// VARIABLES DEFINITION
// Store all the config datasets
var superCDS = {};
// Root node string used to concatenate all CDS in superCDS
var rootNode = "";
// Define keywords in key name that defines a password
var keyNames = [];
// Defines if error must include full path of key found
var includePath = true;
// Defines the max number of errors to return
var maxErrorDisplay = 3;
var errorFound = false;
var errors = [];
var description = '';
// Define the path 
var pathSeparator = "/";

// HANDLERS
// Inputs parser and checker
// Input values in object notation
// Checking the assigned metadasets and parse the node name from input values in object notation
if (arg!=null && cds!=null){
  for (var i=0; i<cds.length; i++){
    rootNode = Object.keys(cds[i])[0];
    superCDS[rootNode] = cds[i][rootNode];
  }
  keyNames=objFormat(arg.trim());
} else {
  errorFound=true;
  errors.push("ERROR: No inputs provided! Please provide at least one arg in object notation.");
}

//MAIN
// here we call our function with different search terms
for (var i= 0; i < keyNames.length; i++) {
  if (keyNames[i] !== "") {
    var myValue = getKeyValueByName(superCDS, keyNames[i]);
    // If found, check if these value exists for same key in all instances
    if (myValue != "ERROR: NOT FOUND") { checkKeyValueExists(superCDS, keyNames[i], myValue, ''); }
  }
}

if (errors.length > 0) {
  errorFound = true;
  if (errors.length < maxErrorDisplay) { description = "ERRORS: " + errors.join(', '); }
  else { description = "ERRORS: only first "+maxErrorDisplay+" errors are displayed:" + errors.join(', '); }
} else { description = "Validation passed successfully"; }

return {description: description, result:!errorFound};

// FONCTIONS LIST
// Parse the object notation: check upon against the RegEx format
function objFormat(obj) {
  var matches = ""; 
  var index = "";
  var keysList = [];
  // {"keyNames" : ["key1","key2","key3"]}
  var jsonRegex = /^\{/gm;
  // <keyNames>
  //		<keyName>key1</keyName>
  //		<keyName>key2</keyName>
  //		<keyName>key3</keyName>
  // </keyNames>
  var xmlRegex = /\<.*\>(.*?)<\/.*\>/gm;
  // keyNames:
  //	-key1 
  //	-key2 
  //	-key3
  var yamlRegex = /.*\-(.*?)$/gm;
  // JSON
  if (jsonRegex.test(obj)) {
    matches = JSON.parse(obj);
    keysList = matches.keyNames;
    return keysList;
  }
  // XML
  else if (xmlRegex.test(obj)) {
    matches = Array.from(obj.matchAll(xmlRegex));
    for (index in matches) {keysList.push(matches[index][1]);}
    return keysList;
  }
  // YAML
  else if (yamlRegex.test(obj)) {
    matches = Array.from(obj.matchAll(yamlRegex));
    for (index in matches) {keysList.push(matches[index][1]);}
    return keysList;
   }
// Unexpected Inputs
  else {
    errorFound=true;
    errors.push("ERROR: Inputs unexpected!, the arg object must contains an array of strings");
  }
}

function getKeyValueByName(mds, keyName) {
  var value = "ERROR: NOT FOUND";
  for (var item in mds) {
    // check if the key has a value or points to an object
    if  (typeof (mds[item]) === "object") {
      // if value is an object call recursively the function to search this subset of the object
      value = getKeyValueByName(mds[item], keyName);
      if (value != "ERROR: NOT FOUND") { return value; }
    } else {
      // check if the key equals to the search term
      if (item === keyName ) { return mds[item]; }
    }
  }
  return value;
}

function checkKeyValueExists (mds, keyname, keyvalue, path) {
  for (var item in mds) {
    if (errors.length >= maxErrorDisplay) { break; }
    // check if the key has a value or points to an object
    if  (typeof (mds[item]) === "object") {
      // if value is an object call recursively the function to search this subset of the object
      checkKeyValueExists(mds[item], keyname, keyvalue, path+item+pathSeparator);
    } else {
      // check if the key & value equals to the search term
      if (item === keyname && mds[item] === keyvalue) { errors.push("Key: "+path+item+" contains same value: "+keyvalue); }
    }
  }
}

File: duplicateKeyValue.py
# description: Check if there is no duplicate key value between conf and deployed instances
# duplicateKeyValue.py
#
# Inputs are: the keys list to search for duplicates
#    Input type: an object arg containing a string
# Outputs are: the list of keys which are duplicated
#    Output type: config datasets
#
# Creator: Cyrille
# Maintainer:
# Version:   0.9
# Support: Sweagle version >= 3.11

# IMPORT MODULES
# JSON module
import json
# RegEx module
#import re
# XML module
#import xml.etree.ElementTree as xml
# YAML module
#import yaml

# FUNCTIONS LIST
# Parse the object notation: check upon against the RegEx format
def objFormat(obj):
  keysList = list()
  # {"KeysList" : ["key1","key2","key3"]}
  # JSON format
  # <KeysList>Value</KeysList>
  # XML format
  # ---
  # KeysList: Value
  # YAML format
  if obj[0]=='{':
    # JSON
    jsonObj=json.loads(obj)
    keysList=jsonObj["KeysList"]
    return keysList
  # XML
  elif obj[0]=='<':
    #xmlObj=xml.fromstring(obj)
    #valueToCheck=xmlObj[0].text
    return keysList
  # YAML
  elif obj[0]=='-':
    #yamlObj=yaml.load(obj)
    #valueToCheck=yamlObj["KeysList"]
    return keysList
  else:
    global errors
    errors.append("ERROR: Inputs unexpected!, please provide an object notation (arg). Inputs variables list (args[]) is deprecated.")

# Function to find keys with duplicate values in CDS
def getDuplicateKeyValue(dataset, keyName):
  global errors
  value = "ERROR: NOT FOUND"
  flipped = {}
  # If the number of errors found is reached
  if len(errors) <= maxErrorDisplay:
    if keyName in dataset:
      # Once the key name found into the CDS, search all its values by flipping the dictionary
      for k, v in dataset.items():
        if v not in flipped:
          flipped[v] = [k]
        else:
          # Duplicated K/V found and return it
          flipped[v].append(k)
          errors.append("Key: "+keyName+" contains same value: "+str(flipped))  
    # Search into the CDS the key name
    for k, v in dataset.items():
      # Recursive search in the CDS   
      if isinstance(v,dict):
        item = getDuplicateKeyValue(v, keyName)
        if item is not None: return item
      else:
        return value

# HANDLERS
# Inputs parser and checker
def handlers(arg):
  if arg!=None and arg!="":
    # Input values in object notation
    # Checking the assigned config datasets and parse the node name from input values in object notation
    return objFormat(arg.strip())
  else:
    global errors
    errors.append("ERROR: No inputs provided! Please provide at least one cds and one arg in object notation.")
    # If no input is provided then return main cds (for retro compatibility)
    return cds[0]

# VARIABLES DEFINITION
# Store all the config datasets
superCDS = {}
# Root node string used to concatenate all CDS in superCDS
rootNode = ""
# Define keywords in key name that defines a password
keyNames = list()
# Defines if error must include full path of key found
includePath = True
# Errors variables
maxErrorDisplay = 3
errorFound = False
errors = list()
errors_description = ""
# Define the path
pathSeparator = "/"
# Define my value
myValue = ""

# MAIN
# If the key names have been correctly parsed without any error detected so far!
# CHECK IF ARG IS WELL DEFINED AND RETRIVE ITS CONTENT
keyNames=handlers(arg)
# MAIN ROUTINE
# here we call our function with different search terms
if keyNames is not None and errorFound is not True:
  global errors
  for i in range(0, len(cds)):
    superCDS = cds[i]
  # For each key name in the list
  for k in range(0, len(keyNames)):
    myValue=getDuplicateKeyValue(superCDS, keyNames[k])
    if myValue is "ERROR: NOT FOUND":
        errors.append(myValue)
  # Return the validation status
  if len(errors) > 0:
    errorFound = True
    if len(errors) < maxErrorDisplay:
      errors_description = "ERRORS: " +', '.join(errors)
    else:
      errors_description = "ERRORS: only first "+maxErrorDisplay+" errors are displayed:" +', '.join(errors)
else:
   errors_description = "Validation passed successfully"
return {"description":errors_description, "result":not errorFound}